fw/hpack.c

```diff
@@ -3930,23 +3994,128 @@ tfw_hpack_encode(TfwHttpResp *__restrict resp, TfwStr *__restrict hdr,
 void
 tfw_hpack_set_rbuf_size(TfwHPackETbl *__restrict tbl, unsigned short new_size)
 {
-	if (WARN_ON_ONCE(new_size > HPACK_ENC_TABLE_MAX_SIZE))
-		return;
+	if (new_size > HPACK_ENC_TABLE_MAX_SIZE) {
+		T_WARN("Client requests hpack table size (%hu), which is "
+			"greater than HPACK_ENC_TABLE_MAX_SIZE.", new_size);
+		new_size = HPACK_ENC_TABLE_MAX_SIZE;
+	}
 
 	spin_lock(&tbl->lock);
 
 	T_DBG3("%s: tbl->rb_len=%hu, tbl->size=%hu, tbl->window=%hu,"
 	       " new_size=%hu\n", __func__, tbl->rb_len, tbl->size,
 	       tbl->window, new_size);
 
-	if (tbl->window > new_size) {
+	if (tbl->window == new_size)
+		goto out;
+
+	/*
+	 * RFC7541#section-4.2:
+	 * Multiple updates to the maximum table size can occur between the
+	 * transmission of two header blocks.  In the case that this size is
+	 * changed more than once in this interval, the smallest maximum table
+	 * size that occurs in that interval MUST be signaled in a dynamic table
+	 * size update.
+	 */
+	if (likely(!tbl->wnd_changed) || unlikely(!tbl->window) ||
+	    new_size < tbl->window) {
 		if (tbl->size > new_size)
 			tfw_hpack_rbuf_calc(tbl, new_size, NULL,
 					    (TfwHPackETblIter *)tbl);
 		WARN_ON_ONCE(tbl->rb_len > tbl->size);
 
 		tbl->window = new_size;
+		tbl->wnd_changed = 1;
+	}
+
+out:
+	spin_unlock(&tbl->lock);
+}
+
+#define __TFW_HPACK_ENC_TBL_SIZE_CHANGED_DECLARE_VARS(resp)		\
+	struct sk_buff __maybe_unused **skb_head = &resp->msg.skb_head;	\
+	unsigned long *flags = resp->flags;				\
+	TfwH2Ctx *ctx = tfw_h2_context(resp->req->conn);		\
+	TfwHPackETbl *tbl = &ctx->hpack.enc_tbl;			\
+	TfwHttpTransIter *mit = &resp->mit;				\
+	TfwStr new_size = {};						\
+	int r = 0;
+
+#define __TWF_HPACK_ENC_TBL_SIZE_CHECK_WR_RESULT(r)			\
+	if (likely(!r)) {						\
+		__set_bit(TFW_HTTP_B_H2_ENC_TBL_SIZE_CONTAINED, flags);	\
+		resp->window = tbl->window;				\
+		*bytes_written = new_size.len;				\
+	}
+
+static inline void
+tfw_hpack_enc_tbl_size_write(TfwHPackETbl *__restrict tbl, TfwStr *new_size)
+{
+	TfwHPackInt tmp;
+
+	write_int(tbl->window, 0x1F, 0x20, &tmp);
+	new_size->data = tmp.buf;
```

Comment:
This looks like storing a pointer to the stack frame that had been destroyed to me.

Reply:
Yes it's my bad mistake!